python/pyarrow/tests/test_ipc.py

```diff
@@ -126,6 +128,83 @@ def test_read_all(self):
         assert result.equals(expected)
 
 
+class TestSocket(MessagingTest, unittest.TestCase):
+
+    class EchoServer(threading.Thread):
+
+        def init(self, do_read_all):
+            self._sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            self._sock.bind(('127.0.0.1', 0))
+            self._sock.listen(1)
+            host, port = self._sock.getsockname()
+            self._do_read_all = do_read_all
+            self._schema = None
+            self._batches = []
+            self._table = None
+            return port
+
+        def run(self):
+            connection, client_address = self._sock.accept()
+            try:
+                source = connection.makefile(mode='rb')
+                reader = pa.StreamReader(source)
+                self._schema = reader.schema
+                if self._do_read_all:
+                    self._table = reader.read_all()
+                else:
+                    for i, batch in enumerate(reader):
+                        self._batches.append(batch)
+            finally:
+                connection.close()
+
+        def get_result(self):
+            return(self._schema, self._table if self._do_read_all else self._batches)
+
+    def setUp(self):
+        # NOTE: must start and stop server in test
+        pass
+
+    def startClientServer(self, do_read_all):
+        self._server = TestSocket.EchoServer()
+        port = self._server.init(do_read_all)
+        self._server.start()
+        self._sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self._sock.connect(('127.0.0.1', port))
+        self.sink = self._get_sink()
+
+    def stopClientServer(self):
+        import struct
+        self.sink.write(struct.pack('i', 0))
```

Comment:
Sure. I am not sure if it will be more complicated than writing a 4-byte 0 in https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/writer.cc#L623. If you'd rather do that in a separate patch I can take a look after this goes in